pkg/controller/podautoscaler/horizontal.go

```diff
@@ -200,160 +200,238 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
 
 	statuses = make([]autoscalingv2.MetricStatus, len(metricSpecs))
 
+	if scale.Status.Selector == "" {
+		errMsg := "selector is required"
+		a.eventRecorder.Event(hpa, v1.EventTypeWarning, "SelectorRequired", errMsg)
+		setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "InvalidSelector", "the HPA target's scale is missing a selector")
+		return 0, "", nil, time.Time{}, fmt.Errorf(errMsg)
+	}
+
+	selector, err := labels.Parse(scale.Status.Selector)
+	if err != nil {
+		errMsg := fmt.Sprintf("couldn't convert selector into a corresponding internal selector object: %v", err)
+		a.eventRecorder.Event(hpa, v1.EventTypeWarning, "InvalidSelector", errMsg)
+		setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "InvalidSelector", errMsg)
+		return 0, "", nil, time.Time{}, fmt.Errorf(errMsg)
+	}
+
+	var validMetricCondition autoscalingv2.HorizontalPodAutoscalerCondition
+
+	invalidMetricsCount := 0
+	var invalidMetricError error
+	var invalidMetricCondition autoscalingv2.HorizontalPodAutoscalerCondition
+
 	for i, metricSpec := range metricSpecs {
-		if scale.Status.Selector == "" {
-			errMsg := "selector is required"
-			a.eventRecorder.Event(hpa, v1.EventTypeWarning, "SelectorRequired", errMsg)
-			setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "InvalidSelector", "the HPA target's scale is missing a selector")
-			return 0, "", nil, time.Time{}, fmt.Errorf(errMsg)
-		}
+		replicaCountProposal, metricNameProposal, status, timestampProposal, condition, err := a.computeReplicasForMetric(hpa, metricSpec, currentReplicas, selector)
+		statuses[i] = status
 
-		selector, err := labels.Parse(scale.Status.Selector)
 		if err != nil {
-			errMsg := fmt.Sprintf("couldn't convert selector into a corresponding internal selector object: %v", err)
-			a.eventRecorder.Event(hpa, v1.EventTypeWarning, "InvalidSelector", errMsg)
-			setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "InvalidSelector", errMsg)
-			return 0, "", nil, time.Time{}, fmt.Errorf(errMsg)
+			if invalidMetricsCount <= 0 {
+				invalidMetricCondition = condition
+				invalidMetricError = err
+			}
+			invalidMetricsCount++
+		}
+		if err == nil && (replicas == 0 || replicaCountProposal > replicas) {
+			timestamp = timestampProposal
+			replicas = replicaCountProposal
+			metric = metricNameProposal
+			validMetricCondition = condition
 		}
+	}
 
-		var replicaCountProposal int32
-		var utilizationProposal int64
-		var timestampProposal time.Time
-		var metricNameProposal string
+	// If all metrics are invalid or some are invalid and we would scale down,
+	// return error and set condition on hpa based on first invalid metric.
+	if invalidMetricsCount >= len(metricSpecs) || (invalidMetricsCount > 0 && replicas < currentReplicas) {
```

Comment:
It feels like the naming of `replicas` here is a bit misleading now (although the name still makes sense at return time) as it's only the current proposed new replica count at the time of this comparison.

It took me a few scans to figure out what it actually corresponded to at the point where this comparison is done.

Reply:
Must work so much better.